server/src/main/java/org/apache/iotdb/db/query/control/QueryResourceManager.java

```diff
@@ -85,20 +109,49 @@ public void registerTempExternalSortFile(long queryId,
     externalSortFileMap.computeIfAbsent(queryId, x -> new ArrayList<>()).add(deserializer);
   }
 
-
   public QueryDataSource getQueryDataSource(Path selectedPath,
       QueryContext context, Filter filter) throws StorageEngineException, QueryProcessException {
 
     SingleSeriesExpression singleSeriesExpression = new SingleSeriesExpression(selectedPath,
         filter);
-    return StorageEngine.getInstance().query(singleSeriesExpression, context, filePathsManager);
+    QueryDataSource queryDataSource = StorageEngine.getInstance()
+        .query(singleSeriesExpression, context, filePathsManager);
+    // calculate the distinct number of seq and unseq tsfiles
+    if (config.isEnablePerformanceTracing()) {
+      Set<TsFileResource> seqFileNum = seqFileNumMap.get(context.getQueryId());
+      Set<TsFileResource> unseqFileNum = unseqFileNumMap.get(context.getQueryId());
+      if (seqFileNum == null) {
+        seqFileNumMap.put(context.getQueryId(), new HashSet<>(queryDataSource.getSeqResources()));
+      } else {
+        seqFileNum.addAll(queryDataSource.getSeqResources());
+      }
+      if (unseqFileNum == null) {
+        unseqFileNumMap.put(context.getQueryId(), new HashSet<>(queryDataSource.getUnseqResources()));
+      } else {
+        unseqFileNum.addAll(queryDataSource.getUnseqResources());
+      }
+    }
+    return queryDataSource;
   }
 
   /**
    * Whenever the jdbc request is closed normally or abnormally, this method must be invoked. All
    * query tokens created by this jdbc request must be cleared.
    */
   public void endQuery(long queryId) throws StorageEngineException {
+    try {
+      if (config.isEnablePerformanceTracing() && chunkNumMap.get(queryId) != null) {

```

Comment:
Not all query unpacks the TimeseriesMetadata, we only unpack it when needed. So the chunkNumMap may do not contain the queryId, you'd better check the seqFileNum or check one by one

Reply:
Fixed.